core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionRepository.java

```diff
@@ -182,6 +182,9 @@ private FunctionBuilder getFunctionBuilder(
     if (isCastFunction(functionName) || sourceTypes.equals(targetTypes)) {
       return funcBuilder;
     }
+    if (functionName.equals(BuiltinFunctionName.CONCAT.getName())) {
+      return castArguments(sourceTypes, sourceTypes, funcBuilder);
+    }
```

Comment:
As `concat()` has variable number of args and targetTypes will always be 0 (as the function signature is not fixed), failure will occur. So, in this case `sourceTypes` are passed instead of `targetTypes` to address that.

Reply:
I think `castArguments` shouldn't be used for `concat`. `getFunctionBuilder` should follow `sourceTypes.equals(targetTypes)` condition (line 182) and return `funcBuilder` as is (line 183).
Even we don't know amount of arguments, they are always STRINGS and user should explicitly do cast if needed. Automatic cast to `STRING` in `concat` could be a follow-up feature.

Reply:
The issue is that `sourceTypes.equals(targetTypes)` will always be `false` as `targetTypes` contains 0 elements, whereas `sourceTypes` can contain variable number of elements.

Reply:
We need to tweak that in the brand new `VarargFunctionResolver`. First (maybe wrong) idea that I got:
1. Make new function type, like you did for `position` function
2. Make it argumentless for repository, like `concat()`
3. Make signature without arguments
4. Let new resolver properly pass arguments to the function

Another thought to replace
```java
if (functionName.equals(BuiltinFunctionName.CONCAT.getName())) {
  ...
}
```
with
```java
if (functionResolverMap.get(functionName) instanceof VarargFunctionResolver) {
  ...
}
```

Reply:
Thank you very much for the feedback, Yury! Will try implementing the suggested changes.